### DISCUSSION TITLE
Refactor examples

### DISCUSSION THREAD

**Discussion by winksaville** [schteve/nucleo-f446re#7] [Ideas]
I'd like to suggest refactoring the examples so they are more like a real project, see [microbit-bsp/examples](https://github.com/lulf/microbit-bsp/tree/main/examples).

A downside of this is that to run the examples you need to:
```
cd examples/xxx
cargo run
```
instead of:
```
cargo run --example xxx
```
This probably could be mitigated mabe using xtask, but that adds complexity of it's own :)

**Comment by schteve** [comment]
Hmm, it's an interesting idea. My initial reaction is that these examples are too simple to bother with adding this complexity. They all share the same setup (`Cargo.toml`, `memory.x`, `build.rs`, GDB scripts, etc), so it reduces a lot of duplication, and it's nice to just use the built-in `cargo run --example` to access them (both for development and for users). But, certainly this approach lacks the complete project feel, and the approach you're suggesting would make it trivial for the user to create their own project by copy/pasting. Actually, I'm not sure if the examples you linked are 100% of a project? I don't see duplicates of `.cargo/config.toml` or `rust-toolchain.toml` so maybe the user would need to add a few files anyway. So, I'm not completely sure how I feel.

**Reply by winksaville** [reply]
I learn by reading, running examples and modifying examples so it's my opinion there should be many examples, i.e. you can't have to many. Also, they should range in complexity to trivial "blink" and "hello, world" to relatively complex. What is your opinion  on the number and complexity of examples?

So regarding my refactoring suggestion, if your feeling is a few examples are all that's needed for testing and validation then what you have is good enough. But if you expect expanding the number and complexity then I'd say refactoring is desirable, IMHO :)

**Reply by schteve** [reply]
For me examples are great, but they should demonstrate something new or interesting about the project. If they repeat things that have already been shown then it quickly devolves into clutter. This crate isn't very large so the examples I have already cover a lot, and I would worry that a complex example would just end up focusing on the other stuff instead of this project. Not to say that I'm against more examples - quite the contrary - but I'd lean towards quality over quantity.

One thing I've seen done elsewhere is to have links to projects that incorporate the crate, rather than contain them as examples, which lets the other projects grow and maintain on their own but showcases what can be done. I'd probably be more open to that for stuff on the complex side.

**Comment by winksaville** [comment]
Fair enough, pointing at other code outside of the crate is a good idea as
they would be "complete', which is what I think is very desirable. A
problem I see is you have no control and what might be a good initially
could change over time. I would guess you had a reason to create this
crate, have you created any thing that might make good example code?

**Reply by schteve** [reply]
I don't have anything that I would care to show. I wouldn't be too worried about not having control over the 3rd party projects, the biggest risk would be those projects falling out of date and misleading people if this crate evolved (think: reading a tutorial that is for an older version of something and not realizing it, then getting frustrated when it doesn't work as expected). But, I don't expect that to be a problem for this crate :-)

**Comment by winksaville** [comment]
Ok, txs for the discussion!
